File: apps/restate/src/services/storage-cleanup.ts
import * as restate from "@restatedev/restate-sdk-cloudflare-workers/fetch";
import { serde } from "@restatedev/restate-sdk-zod";
import { z } from "zod";

import { type Env } from "../env";
import { deleteFile, listAllFiles } from "../supabase";

const CleanupInput = z.object({
  cutoffHours: z.number().min(1).default(24),
});

export type CleanupInputType = z.infer<typeof CleanupInput>;

const CleanupResult = z.object({
  deletedCount: z.number(),
  failedCount: z.number(),
  totalScanned: z.number(),
  errors: z.array(z.string()),
});

export type CleanupResultType = z.infer<typeof CleanupResult>;

export const storageCleanup = restate.service({
  name: "StorageCleanup",
  handlers: {
    cleanupOldFiles: restate.handlers.handler(
      { input: serde.zod(CleanupInput) },
      async (
        ctx: restate.Context,
        input: CleanupInputType,
      ): Promise<CleanupResultType> => {
        const env = ctx.request().extraArgs[0] as Env;
        const cutoffMs = input.cutoffHours * 60 * 60 * 1000;
        const cutoffDate = new Date(Date.now() - cutoffMs);

        const files = await ctx.run("list-files", () => listAllFiles(env));

        let deletedCount = 0;
        let failedCount = 0;
        const errors: string[] = [];

        for (const file of files) {
          const fileDate = new Date(file.created_at);
          if (fileDate < cutoffDate) {
            const filePath = file.name;
            try {
              await ctx.run(`delete-${filePath}`, () =>
                deleteFile(env, filePath),
              );
              deletedCount++;
            } catch (err) {
              failedCount++;
              const errorMsg =
                err instanceof Error ? err.message : "Unknown error";
              errors.push(`Failed to delete ${filePath}: ${errorMsg}`);
              if (errors.length >= 10) {
                errors.push("... (truncated, too many errors)");
                break;
              }
            }
          }
        }

        return {
          deletedCount,
          failedCount,
          totalScanned: files.length,
          errors,
        };
      },
    ),
  },
});

export type StorageCleanup = typeof storageCleanup;

File: apps/restate/src/services/storage_cleanup.rs
use restate_sdk::prelude::*;
use restate_sdk::serde::Json;
use serde::{Deserialize, Serialize};

use crate::env::Env;
use crate::supabase;

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CleanupInput {
    #[serde(default = "default_cutoff_hours")]
    pub cutoff_hours: u64,
}

fn default_cutoff_hours() -> u64 {
    24
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CleanupResult {
    pub deleted_count: u64,
    pub failed_count: u64,
    pub total_scanned: u64,
    pub errors: Vec<String>,
}

#[restate_sdk::service]
pub trait StorageCleanup {
    #[name = "cleanupOldFiles"]
    async fn cleanup_old_files(
        input: Json<CleanupInput>,
    ) -> Result<Json<CleanupResult>, HandlerError>;
}

pub struct StorageCleanupImpl {
    env: &'static Env,
    client: reqwest::Client,
}

impl StorageCleanupImpl {
    pub fn new(env: &'static Env) -> Self {
        Self {
            env,
            client: reqwest::Client::new(),
        }
    }
}

impl StorageCleanup for StorageCleanupImpl {
    async fn cleanup_old_files(
        &self,
        ctx: Context<'_>,
        input: Json<CleanupInput>,
    ) -> Result<Json<CleanupResult>, HandlerError> {
        let input = input.into_inner();
        let cutoff_ms = input.cutoff_hours * 60 * 60 * 1000;
        let now_ms = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;
        let cutoff_time = now_ms - cutoff_ms;

        let env = self.env;
        let client = self.client.clone();
        let Json(files): Json<Vec<supabase::StorageFile>> = ctx
            .run(|| async move {
                supabase::list_all_files(&client, env)
                    .await
                    .map(Json)
                    .map_err(|e| TerminalError::new(format!("Failed to list files: {e}")).into())
            })
            .await?;

        let total_scanned = files.len() as u64;
        let mut deleted_count = 0u64;
        let mut failed_count = 0u64;
        let mut errors = Vec::new();

        for file in &files {
            let file_time = chrono::DateTime::parse_from_rfc3339(&file.created_at)
                .map(|dt| dt.timestamp_millis() as u64)
                .unwrap_or(0);

            if file_time < cutoff_time {
                let file_path = file.name.clone();
                let env = self.env;
                let client = self.client.clone();
                let fp = file_path.clone();
                let delete_result: Result<(), TerminalError> = ctx
                    .run(|| async move {
                        supabase::delete_file(&client, env, &fp)
                            .await
                            .map_err(|e| TerminalError::new(e.to_string()).into())
                    })
                    .await;

                match delete_result {
                    Ok(()) => deleted_count += 1,
                    Err(err) => {
                        failed_count += 1;
                        errors.push(format!("Failed to delete {file_path}: {err}"));
                        if errors.len() >= 10 {
                            errors.push("... (truncated, too many errors)".to_string());
                            break;
                        }
                    }
                }
            }
        }

        Ok(Json(CleanupResult {
            deleted_count,
            failed_count,
            total_scanned,
            errors,
        }))
    }
}
